compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -616,10 +616,34 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
             _ if sp == expr.span && !is_macro => {
                 if let Some(steps) = self.deref_steps(checked_ty, expected) {
+                    // Look through blocks
+                    let mut expr = expr;
+                    while let hir::ExprKind::Block(hir::Block { expr: Some(expr_), .. }, _) =
+                        expr.kind
+                    {
+                        expr = expr_;
+                    }
```

Comment:
We have a method called `peel_refs`, we could add `peel_blocks` to `Expr` as well... 🤔

Reply:
Done.